### ISSUE TITLE
[SPIR-V] Calling `var[0].member` as opposed to `var.operator[](0).member` made the compiler go SIGSEGV

### ISSUE THREAD

**Issue by devshgraphicsprogramming** [microsoft/DirectXShaderCompiler#5638]
**Description**

Near the end of the shader there's a 
```cpp
// the following cause us to SIGSEGV
    member_ref_t<float[1],8,12> pB = pLineStyle[0].normalizedStipple;
```
comment this assignment out and it will no longer crash.

the funny thing is that I've coded `deref` and `operator[](0)` to work exactly the same, and I think that the problem might be in `operator[]`.

P.S. @natevm this works around all your issues with BDA, except for the atomics

**Steps to Reproduce**

Version with template + macro debauchery:
https://godbolt.org/z/4Gxc7rjoz

Verison with template and the macros already preprocessed:
https://godbolt.org/z/Pa119ej4n

**Actual Behavior**

Compiler Crashes without diagnostics

**Environment**
- DXC version trunk/Godbolt
- Host Operating System Linux/Godbolt

**Comment by devshgraphicsprogramming** [comment]
OK I found a funny thing, without `-spirv` I get following error about `return this;`
```
error: cannot compile this aggregate expression yet
        return this;
               ^~~~
```
but I think thats a separate bug altogether

**Comment by devshgraphicsprogramming** [comment]
ok but it definitely doesn't crash when emitting DXIL with `RWByteAddressBuffer` in place of BDA:
https://godbolt.org/z/KesGq3z44

but as soon as you add `-spirv` to the commandline it crashes

**Comment by devshgraphicsprogramming** [comment]
Narrowed down the problem, this makes it compile fine
````cpp
    member_ref_t<uint32_t[1],8,12> pB = pLineStyle.operator[](0).normalizedStipple;
```

so this means that the problem is with `operator[]` calling

**Comment by cassiebeckley** [comment]
I think there's at least three separate segfaults I've encountered while trying to create a minimal reproducible example for this. I've got a PR out to fix one of them, I'll look at the others when I have time.

**Comment by cassiebeckley** [comment]
I came back to this to try to find the root cause, and I have found a smaller case that I think is touching on something close to what's going on here: https://godbolt.org/z/hnbznvz66. In this example, the fields of the base class are being ignored and not being initialized correctly. It might be an entirely separate issue, but the original case posted above fails in the debug build with an assert about initializers not being handled correctly, so I'll try fixing it first.

**Comment by cassiebeckley** [comment]
As mentioned before, this issue highlights multiple separate bugs in DXC. It should be fixed once #5894 and #5896 have landed.

In the future, it will help us fix bug reports faster if you can narrow down compiler bugs to a minimal reproducible example.

**Comment by devshgraphicsprogramming** [comment]
> As mentioned before, this issue highlights multiple separate bugs in DXC. It should be fixed once #5894 and #5896 have landed.
> 
> In the future, it will help us fix bug reports faster if you can narrow down compiler bugs to a minimal reproducible example.

I completely understand that, however with this one I've had to clue how to narrow down.

**Comment by cassiebeckley** [comment]
Fixed by #5894 and #5896.
